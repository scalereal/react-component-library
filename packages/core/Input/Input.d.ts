import { FC } from 'react';
import { InputProps } from './types';
import { WithStyle } from "../../utils";
export declare const Input: FC<InputProps> & WithStyle;
export default Input;
