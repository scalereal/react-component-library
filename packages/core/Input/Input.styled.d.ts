import { InputProps } from './types';
export declare const OuterWrapper: import("styled-components").StyledComponent<"div", any, InputProps, never>;
export declare const Error: import("styled-components").StyledComponent<"span", any, InputProps, never>;
export declare const InputWrapper: import("styled-components").StyledComponent<"div", any, InputProps, never>;
export declare const Input: import("styled-components").StyledComponent<"input", any, InputProps, never>;
export declare const Preffix: import("styled-components").StyledComponent<"span", any, InputProps, never>;
export declare const Suffix: import("styled-components").StyledComponent<"span", any, InputProps, never>;
