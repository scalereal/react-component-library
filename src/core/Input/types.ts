import React from "react"
export interface InputProps extends React.HTMLProps<HTMLInputElement>{
    /** Takes full width of the parent component */
    fullWidth?: boolean;
    /** Prefix  */
    prefix?: string;
    /** Suffix */
    suffix?: string;
    /** Disable Input */
    disabled?: boolean;
    /** To be used when it is required in any form */
    required?: boolean;
}
