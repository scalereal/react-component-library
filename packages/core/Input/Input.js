var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import Text from '../Text/Text';
import * as Styled from './Input.styled';
export var Input = React.memo(React.forwardRef(function (props, ref) {
    var fullWidth = props.fullWidth, required = props.required, preffix = props.preffix, suffix = props.suffix, borderColor = props.borderColor, textColor = props.textColor, disabledColor = props.disabledColor, backgroundColor = props.backgroundColor, disabled = props.disabled, placeHolderColor = props.placeHolderColor, preffixBackgroundColor = props.preffixBackgroundColor, suffixBackgroundColor = props.suffixBackgroundColor, hasError = props.hasError, errorText = props.errorText, preffixColor = props.preffixColor, suffixColor = props.suffixColor, restProps = __rest(props, ["fullWidth", "required", "preffix", "suffix", "borderColor", "textColor", "disabledColor", "backgroundColor", "disabled", "placeHolderColor", "preffixBackgroundColor", "suffixBackgroundColor", "hasError", "errorText", "preffixColor", "suffixColor"]), id = props.id || 'input';
    return (React.createElement(Styled.OuterWrapper, __assign({}, restProps),
        hasError && React.createElement(Text, { className: "error" }, errorText),
        React.createElement(Styled.InputWrapper, { fullWidth: fullWidth, borderColor: borderColor, textColor: textColor, hasError: hasError },
            preffix && React.createElement(Styled.Preffix, { preffixBackgroundColor: preffixBackgroundColor, preffixColor: preffixColor, hasError: hasError }, preffix),
            React.createElement(Styled.Input, __assign({ id: id, ref: ref, required: required, fullWidth: fullWidth, disabledColor: disabledColor, backgroundColor: backgroundColor, placeHolderColor: placeHolderColor, disabled: disabled }, restProps)),
            suffix && React.createElement(Styled.Suffix, { suffixBackgroundColor: suffixBackgroundColor, suffixColor: suffixColor, hasError: hasError }, suffix))));
}));
Input.displayName = 'Input';
Input.Style = Styled.Input;
Input.defaultProps = {
    type: 'text',
    disabled: false,
    fullWidth: false,
    required: false,
};
export default Input;
