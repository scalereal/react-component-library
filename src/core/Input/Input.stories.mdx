import { Input } from './Input';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Input"
    component={Input}
    argTypes={{
        textColor: {
            control: 'color'
        },
        borderColor: {
            control: 'color'
        },
        disabledColor: {
            control: 'color'
        },
        backgroundColor: {
            control: 'color'
        },
        placeHolderColor: {
            control: 'color'
        },
        preffixBackgroundColor: {
            control: 'color'
        },
        suffixBackgroundColor: {
            control: 'color'
        },
        preffixColor: {
            control: 'color'
        },
        suffixColor: {
            control: 'color'
        }
    }}
/>

# Input

Displays a Input

export const Template = ({
    suffix,
    preffix,
    required,
    disabled,
    fullWidth,
    textColor,
    borderColor,
    disabledColor,
    backgroundColor,
    placeHolderColor,
    preffixBackgroundColor,
    suffixBackgroundColor,
    hasError,
    errorText,
    ...args
}) => {
    return (
        <Input
            id="input"
            type="email"
            fullWidth={fullWidth}
            disabled={disabled}
            required={required}
            placeholder="Enter Email Address"
            preffix={preffix}
            suffix={suffix}
            textColor={textColor}
            borderColor={borderColor}
            disabledColor={disabledColor}
            backgroundColor={backgroundColor}
            placeHolderColor={placeHolderColor}
            preffixBackgroundColor={preffixBackgroundColor}
            suffixBackgroundColor={suffixBackgroundColor}
            hasError={hasError}
            errorText={errorText}
            {...args}
        ></Input>
    );
};

<Preview>
    <Story name="Input">{Template.bind({})}</Story>
</Preview>

### Color Variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <div>
        <Input
            id="input"
            type="email"
            placeholder="Enter 
            Email Address"
            placeHolderColor="blue"
            borderColor="blue"
            backgroundColor="rgb(132 130 130 / 20%)"
        />
        <Input
            id="input"
            type="email"
            placeholder="Enter Email Address"
            placeHolderColor="red"
            borderColor="red"
            backgroundColor="rgb(132 130 130 / 20%)"
        />
    </div>
</Preview>

### prefix Suffix

Displays prefix/suffix text or icon.

<Preview withToolbar>
    <div>
        <Input id="input" type="email" placeholder="Enter Email Address" prefix="icon" />
        <Input id="input" type="email" placeholder="Enter Email Address" suffix="icon" />
    </div>
</Preview>

### Error

Displays error if hasError and errorText="Error Message" props are passed

<Preview withToolbar>
    <div>
        <Input id="input" type="email" placeholder="Enter Email Address" hasError errorText="Error" />
    </div>
</Preview>

### Disabled

Disabled Input

<Preview withToolbar>
    <div>
        <Input id="input" type="email" placeholder="Enter Email Address" disabled />
    </div>
</Preview>

### Props

<Props story="Input" />
