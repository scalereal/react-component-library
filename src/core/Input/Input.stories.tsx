import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import Input from "./Input"

export default {
    title: 'Components/Core/Input',
    component: Input,
    argTypes: {
        textColor: {
            control: 'color'
        },
        borderColor: {
            control: 'color'
        },
        disabledColor: {
            control: 'color'
        },
        backgroundColor: {
            control: 'color'
        },
        placeHolderColor: {
            control: 'color'
        },
        preffixBackgroundColor: {
            control: 'color'
        },
        suffixBackgroundColor: {
            control: 'color'
        }
    }
} as Meta;

const Template: Story = ({
    suffix,
    prefix,
    required,
    disabled,
    fullWidth,
    textColor,
    borderColor,
    disabledColor,
    backgroundColor,
    placeHolderColor,
    preffixBackgroundColor,
    suffixBackgroundColor,
    ...args
}) => {
    return (
        <Input
            id="input"
            type="email"
            fullWidth={fullWidth}
            disabled={disabled}
            required={required}
            placeholder="Enter Email Address"
            prefix={prefix}
            suffix={suffix}
            textColor={textColor}
            borderColor={borderColor}
            disabledColor={disabledColor}
            backgroundColor={backgroundColor}
            placeHolderColor={placeHolderColor}
            preffixBackgroundColor={preffixBackgroundColor}
            suffixBackgroundColor={suffixBackgroundColor}
            {...args}
        ></Input>
    );
};

export const Default = Template.bind({});
Default.args = {

}


