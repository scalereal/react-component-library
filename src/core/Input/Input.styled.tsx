import styled from 'styled-components'
import { InputProps } from './types'
import { defaultTheme } from '../../theme'

export const Wrapper = styled('div')<InputProps>`
    display: flex;
    &&& {
        flex-direction: row;
        align-items: center;
    }
    background-clip: padding-box;
    box-sizing: border-box;
    border: 1px solid ${({ theme, borderColor }) => (borderColor ? borderColor : theme.input.inputBorderColor)};
    border-radius: 5px;
    overflow: hidden;
    width: ${props => (props.fullWidth ? '100%' : 'max-content')};
    height: max-content;
    font-family: ${({ theme }) =>(theme.input.fontFamily)};
    &:focus-within {
        border-color: ${({ theme, borderColor }) => (borderColor ? borderColor : theme.input.inputBorderColor)};
    }
    & > * {
        font-size: 1rem;
        padding: 10px;
        font-family: ${({ theme }) =>(theme.input.fontFamily)};
        color: ${({ theme, textColor }) => (textColor ? textColor : theme.input.inputTextColor)};
    }
`;
Wrapper.defaultProps = {
    theme: defaultTheme
}

export const Input = styled('input')<InputProps>`
    flex: 1;
    outline: 0;
    border: 0;
    cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'inherit')};
    font-family: ${({ theme }) =>(theme.fontFamily)};
    background-color: ${({ disabled, theme, disabledColor, backgroundColor }) =>
        disabled
            ? disabledColor
                ? disabledColor
                : theme.input.inputDisabledColor
            : backgroundColor
            ? backgroundColor
            : theme.input.inputBackgroundColor};
    &::placeholder {
        color: ${({theme, placeHolderColor }) => placeHolderColor ? placeHolderColor : theme.input.inputPlaceHolderColor};
        font-family: ${({ theme }) =>(theme.input.fontFamily)};
    }
    &::-webkit-file-upload-button {
        visibility: hidden;
    }
`;
Input.defaultProps = {
    theme: defaultTheme
};

export const Prefix = styled('span')<InputProps>`
    background-color: ${({ theme, preffixBackgroundColor }) =>
        preffixBackgroundColor ? preffixBackgroundColor : theme.input.inputPreffixBackgroundColor};

    font-family: ${({ theme }) =>(theme.input.fontFamily)};
`;
Prefix.defaultProps = {
    theme: defaultTheme
};

export const Suffix = styled('span')<InputProps>`
    background-color: ${({ theme, suffixBackgroundColor }) =>
        suffixBackgroundColor ? suffixBackgroundColor : theme.input.inputSuffixBackgroundColor};
    
    font-family: ${({ theme }) =>(theme.input.fontFamily)};
`;
Suffix.defaultProps = {
    theme: defaultTheme
};