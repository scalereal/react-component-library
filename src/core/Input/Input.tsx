import React, { FC } from 'react';
import * as Styled from './Input.styled';
import { InputProps, WithStyle } from './types';

export const Input: FC<InputProps> & WithStyle = React.memo(
    React.forwardRef((props, ref) => {
        const { fullWidth, required, prefix, suffix, ...restProps } = props,
            id = props.id || 'input';
        return (
            <Styled.Wrapper fullWidth={fullWidth}>
                {prefix && <Styled.Prefix>{prefix}</Styled.Prefix>}
                <Styled.Input id={id} ref={ref} required={required} fullWidth={fullWidth} {...restProps} />
                {suffix && <Styled.Suffix>{suffix}</Styled.Suffix>}
            </Styled.Wrapper>
        );
    })
);
Input.displayName = 'Input';
Input.Style = Styled.Input;
Input.defaultProps = {
    type: 'text',
    disabled: false,
    fullWidth: false,
    required: false,
};
export default Input
