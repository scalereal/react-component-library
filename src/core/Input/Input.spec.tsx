import { render } from '@testing-library/react';
import React from 'react';
import { Input } from './Input';

describe('Input', () => {
    it('should render correctly with all the props given', () => {
        const { container } = render(
            <Input
                id="email-input"
                type="email"
                fullWidth
                required
                placeholder="Enter Email Address"
                prefix="www."
                suffix=".com"
            />
        );
        expect(container).toMatchSnapshot();
    });

    it('should render correctly when fullWidth & required props are falsy', () => {
        const { container } = render(
            <Input
                type="email"
                placeholder="Enter Email Address"
            />
        );
        expect(container).toMatchSnapshot();
    });

    it('should render correctly when it is disabled', () => {
        const { container } = render(<Input id="email-input" disabled placeholder="Enter Email Address" />);
        expect(container).toMatchSnapshot();
    });
});
