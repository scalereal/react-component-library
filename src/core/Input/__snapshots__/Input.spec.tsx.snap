// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly when fullWidth & required props are falsy 1`] = `
<div>
  <div
    class="sc-bdfBwQ dOTqjN"
  >
    <input
      class="sc-gsTCUz guRvgL"
      id="input"
      placeholder="Enter Email Address"
      type="email"
    />
  </div>
</div>
`;

exports[`Input should render correctly when it is disabled 1`] = `
<div>
  <div
    class="sc-bdfBwQ dOTqjN"
  >
    <input
      class="sc-gsTCUz qdJqK"
      disabled=""
      id="email-input"
      placeholder="Enter Email Address"
      type="text"
    />
  </div>
</div>
`;

exports[`Input should render correctly with all the props given 1`] = `
<div>
  <div
    class="sc-bdfBwQ faCpUx"
  >
    <span
      class="sc-dlfnbm csXxpq"
    >
      www.
    </span>
    <input
      class="sc-gsTCUz guRvgL"
      id="email-input"
      placeholder="Enter Email Address"
      required=""
      type="email"
    />
    <span
      class="sc-hKgILt dMhINk"
    >
      .com
    </span>
  </div>
</div>
`;
